File: src/main/java/com/est/curdsample/app/TaskService.java
package com.est.curdsample.app;

import com.est.curdsample.dao.TaskRepository;
import com.est.curdsample.domain.Task;
import com.est.curdsample.dto.TaskDto;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TaskService {

    private final TaskRepository taskRepository;

    @Transactional
    public TaskDto saveTask(TaskDto taskDto) {
        taskDto.setCode(genCode());

        Task task = Task.of(taskDto);
        taskRepository.save(task);

        return taskDto;
    }

    private Task findById(Long id) {
        return taskRepository.findById(id).orElse(null);
    }

    private Task findByCode(String code) {
        return taskRepository.findByCode(code).orElse(null);
    }

    private String genCode() {
        return UUID.randomUUID().toString();
    }

    public TaskDto getByCode(String code) {
        Task task = findByCode(code);

        return TaskDto.from(task);
    }

    @Transactional
    public String removeByCode(String code) {

        Task task = findByCode(code);

        taskRepository.delete(task);

        return code;
    }

    @Transactional
    public TaskDto update(TaskDto taskdto) {

        Task task = findByCode(taskdto.getCode());

        task.update(taskdto);

        return taskdto;
    }

}

File: src/main/java/com/est/curdsample/app/TaskService.kt
package com.est.curdsample.app

import com.est.curdsample.dao.TaskRepository
import com.est.curdsample.domain.Task
import com.est.curdsample.domain.toDescription
import com.est.curdsample.domain.toDto
import com.est.curdsample.dto.TaskDescription
import com.est.curdsample.dto.TaskDto
import com.est.curdsample.dto.TaskPageDto
import com.est.curdsample.dto.toEntity
import com.est.curdsample.exception.TaskNotFoundException
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Slice
import org.springframework.data.domain.Sort
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional(readOnly = true)
class TaskService(
    private val taskRepository: TaskRepository
) {

    @Transactional
    fun checkTaskByCode(code: String): TaskDto {
        val task = findByCode(code)

        task.updateCheck()

        return task.toDto()
    }

    fun getTaskList(pageNum: Int): TaskPageDto {
        val SIZE = 5

        val pageRequest = PageRequest.of(pageNum, SIZE, Sort.Direction.DESC, "createdAt")
        val tasks: Slice<Task> = taskRepository.findAll(pageRequest)

        return TaskPageDto(
            hasNext = tasks.hasNext(),
            data = tasks.content.map { it.toDto() }
        )
    }

    fun getDescriptionByCode(code: String): TaskDescription {
        return findByCode(code).toDescription()
    }

    fun getTasksDueToToday(): List<TaskDto> {

        return taskRepository.findTenTasksDueToToday().map { it.toDto() }
    }

    @Transactional
    fun saveTask(taskDto: TaskDto): TaskDto {
        taskDto.code = genCode()

        taskRepository.save(taskDto.toEntity())

        return taskDto
    }

    private fun findByCode(code: String): Task {
        return taskRepository.findByCode(code)
            ?: throw TaskNotFoundException()
    }

    private fun genCode(): String {
        return UUID.randomUUID().toString()
    }

    fun getByCode(code: String): TaskDto {
        val task = findByCode(code)

        return task.toDto()
    }

    @Transactional
    fun removeByCode(code: String): String {
        val task = findByCode(code)

        taskRepository.delete(task)

        return code
    }

    @Transactional
    fun update(taskDto: TaskDto): TaskDto {
        val task = findByCode(taskDto.code)

        task.update(taskDto)

        return taskDto
    }
}
